app/src/main/java/com/eis/phoneringer/structure/AppManager.java

```diff
@@ -0,0 +1,99 @@
+package com.eis.phoneringer.structure;
+
+import android.content.Context;
+import android.media.Ringtone;
+import android.os.Build;
+import android.os.Handler;
+import android.os.Looper;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+
+import com.eis.phoneringer.exceptions.WrongPasswordException;
+import com.eis.smslibrary.SMSHandler;
+import com.eis.smslibrary.exceptions.InvalidSMSMessageException;
+import com.eis.smslibrary.exceptions.InvalidTelephoneNumberException;
+import com.eis.smslibrary.listeners.SMSSentListener;
+
+/**
+ * This is a singleton class used to manage a received RingCommand or to send one
+ *
+ * @author Alberto Ursino, Luca Crema, Alessandra Tonin, Marco Mariotto
+ */
+public class AppManager {
+
+    private static final int TIMEOUT_TIME = 30 * 1000; //30 seconds
+
+    /**
+     * Instance of the class that is instantiated in getInstance method
+     */
+    private static AppManager instance = null;
+
+    /**
+     * Private constructor
+     */
+    private AppManager() {
+    }
+
+    /**
+     * @return the AppManager instance
+     */
+    public static AppManager getInstance() {
+        if (instance == null)
+            instance = new AppManager();
+        return instance;
+    }
+
+    /**
+     * If the password of the RingCommand received is valid then play ringtone for fixed amount of time
+     *
+     * @param context     of the application
+     * @param ringCommand received
+     * @param ringtone    A valid ringtone to be played
+     * @throws WrongPasswordException Exception thrown when the password received is not valid
+     */
+    public void onRingCommandReceived(Context context, @NonNull RingCommand ringCommand, final Ringtone ringtone) throws WrongPasswordException {
+        final RingtoneHandler ringtoneHandler = RingtoneHandler.getInstance();
+
+        if (checkPassword(context, ringCommand)) {
+            ringtoneHandler.playRingtone(ringtone);
+            //Timer: the ringtone is playing for TIMEOUT_TIME seconds.
+            new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    ringtoneHandler.stopRingtone(ringtone);
+                }
+            }, TIMEOUT_TIME);
+        } else {
+            throw new WrongPasswordException();
+        }
```

Comment:
A quick suggestion (even though it's not necessary, you might still find this useful)
There's a way of programming where you focus on having the parameters you want.
Basically code like this
```
if(condition){
    //code
}
else {
    return;
}
```
Becomes
```
if( !condition){
    return;
}
//code
```

This way long nested if chains just become linear
```
if( c1){
    if( c2){
        if( c3){
            //code
        }
    }
}
```
becomes
```
if( !c1)
   return;
if( !c2)
    return;
if( !c3)
    return;

//code
```

In your case it would be
```
if ( !checkPassword(context, ringCommand)) throw new WrongPasswordException();

//code inside the if
```

Just a good way to write code in my opinion.

Reply:
Do you mean something like that?

![Untitled](https://user-images.githubusercontent.com/56396708/70151069-5b960b00-16ab-11ea-8d02-2ef62e9d7010.png)

Reply:
I can change if you prefer.
In my opinion you are in right @Cogno-IDU, this _code writing method_ significantly speeds up both the method and, in the long run, the application itself.

Reply:
Nah you don't need to change it. I just thought you didn't know about it, or maybe you didn't pay attention, so I just gave some feedback 😄 .
Also yeah, it's a pretty useful coding style, it even has it's own name. Something like... defensive programming?